telegram/telegram.go

```diff
@@ -0,0 +1,412 @@
+	storageFile := path.Join(d.config.DownloadPath, d.config.ProgressFile)
+	_, err := progress.NewProgress(int64(current), int64(last), storageFile)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func (d *downloader) DownloadFile(entity *tgFile) {
+
+	tool := downloader2.NewDownloader()
+
+	// Remove the exist file.
+	if _, err := os.Stat(entity.filePath); err == nil {
+		if err := os.Remove(entity.filePath); err != nil {
+			log.Fatal(err)
+		}
+	}
+	writer, err := os.Create(entity.filePath)
+	if err != nil {
+		log.Fatal("create file err [%s]  %s", entity.filePath, err)
+	}
+	defer func() { _ = writer.Close() }()
+
+	thread := 1
+	if entity.fileSize/(512*1024) > d.config.Thread*4 {
+		thread = d.config.Thread
+	}
+	// Add startDownloads progress
+	bar := log.NewProgressBar(int64(entity.id), int64(LastId), entity.format+" "+entity.filename, int64(entity.fileSize))
+
+	_, err2 := tool.Download(d.client.API(), entity.documentFile).WithThreads(thread).Stream(d.context, io.MultiWriter(writer, bar))
+	if err2 != nil {
+		log.Fatal(err)
+	}
+}
+
+func (d *downloader) toFile(message tg.MessageClass, dir string) (*tgFile, bool) {
+	if message == nil {
+		return nil, false
+	}
+	msg, ok := message.(*tg.Message)
+	if !ok {
+		return nil, false
+	}
+	if msg.Media == nil {
+		return nil, false
+	}
+	s, ok := msg.Media.(*tg.MessageMediaDocument)
+	if !ok {
+		return nil, false
+	}
+	document := s.Document.(*tg.Document)
+	fileName := ""
+	for _, attribute := range document.Attributes {
+		x, ok := attribute.(*tg.DocumentAttributeFilename)
+		if ok {
+			fileName = x.FileName
+		}
+	}
+	if len(fileName) == 0 {
+		return nil, false
+	}
+	format := spider.Extension(fileName)
+	outFilename := strconv.FormatInt(int64(msg.ID), 10) + "." + strings.ToLower(format)
+	if !d.rename {
+		outFilename = strconv.Itoa(msg.GetID()) + "_" + fileName
+	}
+	saveFilePath := path.Join(dir, rename.EscapeFilename(outFilename))
+	return &tgFile{
+		id:           msg.ID,
+		filename:     fileName,
+		format:       format,
+		fileSize:     document.Size,
+		documentFile: document.AsInputDocumentFileLocation(),
+		filePath:     saveFilePath,
+	}, true
+}
+
+func (d *downloader) formatMatcher(fileName string) bool {
+	for _, f := range d.config.Formats {
+		if strings.HasSuffix(fileName, strings.ToLower(f)) {
+			return true
+		}
+	}
+	return false
+}
+
+func (d *downloader) Fork() {
+	d.wait.Add(1)
+}
+
+func (d *downloader) Done() {
+	d.wait.Done()
+}
+
+func (d *downloader) Join() {
+	d.wait.Wait()
+}
+
+func generatePart(start int, length int, step int) []filePart {
+	eachSize := length / step
+	if eachSize == 0 {
+		eachSize = 1
+	}
+	if length%step > 0 {
+		eachSize = eachSize + 1
+	}
+	jobs := make([]filePart, eachSize)
+
+	for i := range jobs {
+		jobs[i].Index = i
+		if i == 0 {
+			jobs[i].Offset = start
+		} else {
+			jobs[i].Offset = jobs[i-1].Offset + jobs[i-1].Limit
+		}
+		jobs[i].Limit = step
+		//log.Infof("part %d --> %d", jobs[i].Offset, jobs[i].Offset+jobs[i].Limit)
```

Comment:
We can add debug level for verbose log.

Reply:
I'll add the verbose log after your groundwork last night